subsys/nrf_compress/src/lzma.c

```diff
@@ -303,6 +422,260 @@ static int lzma_decompress(void *inst, const uint8_t *input, size_t input_size,
+	    *offset < input_size) {
+		/* If last block, ensure offset matches complete file size */
+		*offset = input_size;
+	}
+
+	if (decoder->dicPos >= decoder->dicHandle->dicBufSize ||
+	    (last_part && input_size == chunk_size)) {
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+		if (cache.invalid) {
+			validate_cache(decoder->dicHandle);
+		}
+#endif
+		*output_size = decoder->dicPos;
+		decoder->dicPos = 0;
+	}
+
+done:
+	return rc;
+}
+
+static DictHandle dict_handle;
+
+DictHandle *LzmaDictionaryOpen(SizeT size)
+{
+	if (dict_handle.isOpened)
+		return &dict_handle;
+
+	if (ext_dict ==  NULL)
+		return NULL;
+
+	size_t dict_size;
+
+	if (ext_dict->open((size_t)size, &dict_size) != 0) {
+		LOG_WRN("Unable to open external dictionary with size %u", size);
+		return NULL;
+	}
+
+	dict_handle.isOpened = True;
+	dict_handle.dicBufSize = dict_size;
+
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+	cache.dict_pos_begin = 0;
+	cache.dict_pos_end = sizeof(cache.data) - 1;
+#endif
+
+	return &dict_handle;
+}
+
+SizeT LzmaDictionaryWrite(DictHandle *handle, SizeT pos, const Byte *data, SizeT len)
+{
+	if (handle != &dict_handle || ext_dict == NULL)
+		return 0;
+
+	SizeT write_len = len;
+
+	if (pos + len > handle->dicBufSize) {
+		write_len = handle->dicBufSize - pos;
+	}
+
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+	if (pos > cache.dict_pos_end || pos < cache.dict_pos_begin) {
+		/*
+		 * Should never happen, lzma operates on dicPos when writing to dictionary,
+		 * which should be aligned with cache.
+		 */
+		return 0;
+	}
+
+	SizeT bytes_written = 0;
+	SizeT cache_pos = pos - cache.dict_pos_begin;
+
+	while (bytes_written != write_len) {
+		SizeT cache_write_len =
+			(write_len - bytes_written) > (sizeof(cache.data) - cache_pos) ?
+				(sizeof(cache.data) - cache_pos) :
+				(write_len - bytes_written);
+
+		memcpy(cache.data + cache_pos, data + bytes_written, cache_write_len);
+		cache.invalid = true;
+
+		bytes_written += cache_write_len;
+		cache_pos += cache_write_len;
+
+		if (cache_pos == sizeof(cache.data)) {
+			/** Cache full, validate it.*/
+			validate_cache(handle);
+			cache_pos = 0;
+		}
+	}
+#else
+	ext_dict->write(pos, data, write_len);
+#endif
+	return write_len;
+}
+
+SizeT LzmaDictionaryRead(DictHandle *handle, SizeT pos, Byte *data, SizeT len)
+{
+	if (handle != &dict_handle || ext_dict == NULL)
+		return 0;
+
+	int read_len = len;
+
+	if (pos + len > handle->dicBufSize) {
+		read_len = handle->dicBufSize - pos;
+	}
+
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+	SizeT bytes_read = 0;
+
+	if (pos + read_len > cache.dict_pos_begin && pos <= cache.dict_pos_end) {
+		/** We have at least some of the requested data in the cache. */
+		SizeT cache_pos;
+		SizeT cache_copy_size;
+
+		if (pos < cache.dict_pos_begin) {
+			/** First part of data is from dictionary... */
+			ext_dict->read(pos, data, cache.dict_pos_begin - pos);
+			bytes_read = cache.dict_pos_begin - pos;
```

Comment:
We should rather use the return value from `ext_dict->read` and then ensure that that it is equal to this value. There doesn't seem to be any error validation that `read` succeeded